src/codemodder/utils/format_string_parser.py

```diff
@@ -0,0 +1,187 @@
+    piece_parts,
+    token_count: int,
+    keys: dict | list,
+) -> (
+    tuple[
+        list[
+            cst.SimpleString
+            | cst.FormattedStringText
+            | PrintfStringExpression
+            | PrintfStringText
+        ],
+        int,
+    ]
+    | None
+):
+    # if it does not contain any %s token we maintain the original
+    if _has_conversion_parts(piece_parts):
+        parsed_parts: list[
+            cst.SimpleString
+            | cst.FormattedStringText
+            | PrintfStringExpression
+            | PrintfStringText
+        ] = []
+        index_count = 0
+        for s in piece_parts:
+            if s:
+                if s.startswith("%"):
+                    # TODO should account for different prefixes when key is extracted
+                    key = extract_mapping_key(s)
+                    match keys:
+                        case dict():
+                            key = extract_mapping_key(s)
+                            if not key:
+                                return None
+                            parsed_parts.append(
+                                PrintfStringExpression(
+                                    origin=piece,
+                                    expression=keys[key],
+                                    key=key,
+                                    index=index_count,
+                                    value=s,
+                                )
+                            )
+                        case list():
+                            parsed_parts.append(
+                                PrintfStringExpression(
+                                    origin=piece,
+                                    expression=keys[token_count],
+                                    key=token_count,
+                                    index=index_count,
+                                    value=s,
+                                )
+                            )
+                    token_count = token_count + 1
+                else:
+                    parsed_parts.append(
+                        PrintfStringText(origin=piece, value=s, index=index_count)
+                    )
+                index_count += len(s)
+        return parsed_parts, token_count
+    return [piece], token_count
+
+
+def expressions_from_replacements(
+    replacements: cst.Tuple | cst.BaseExpression,
+) -> list[cst.BaseExpression]:
+    """
+    Gather all the expressions from a tuple literal.
+    """
+    match replacements:
+        case cst.Tuple():
+            return [e.value for e in replacements.elements]
+    return [replacements]
+
+
+def dict_to_values_dict(
+    expr_dict: dict[cst.BaseExpression, cst.BaseExpression]
+) -> dict[str | cst.BaseExpression, cst.BaseExpression]:
+    return {
+        extract_raw_value(k): v
+        for k, v in expr_dict.items()
+        if isinstance(k, cst.SimpleString | cst.FormattedStringText)
+    }
+
+
+def parse_formatted_string(
```

Comment:
This is pretty heroic, nice work. I could have sworn that there was something in the standard library that helped to tokenize format strings but I'm having trouble digging it up right now.